commercial/app/model/merchandise/books/BookFinder.scala

```diff
@@ -2,67 +2,47 @@ package commercial.model.merchandise.books
-import common.Logging
+import commercial.model.merchandise.Book
+import common.{AkkaAgent, Logging}
 import conf.Configuration
 import conf.switches.Switches.BookLookupSwitch
-import commercial.model.merchandise.Book
 import play.api.libs.json._
 import play.api.libs.oauth.{ConsumerKey, OAuthCalculator, RequestToken}
 import play.api.libs.ws.{WSClient, WSSignatureCalculator}
-import shade.memcached.{Memcached, MemcachedCodecs, Configuration => MemcachedConfiguration}
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
 
 class BookFinder(actorSystem: ActorSystem, magentoService: MagentoService) extends Logging {
 
-  private implicit lazy val executionContext = actorSystem.dispatchers.lookup("akka.actor.memcached")
-
-  private lazy val defaultBookDataCache = new MemcachedBookDataCache(actorSystem)
-
-  def findByIsbn(isbn: String,
-                 cache: BookDataCache = defaultBookDataCache,
-                 lookup: String => Future[Option[JsValue]] = magentoService.findByIsbn):
-  Future[Option[Book]] = {
-
-    def cachedBook(bookData: JsValue): Future[Option[Book]] = {
-      Future.successful {
-        bookData match {
-          case JsNull => None
-          case _ =>
-            bookData.validate[Book] fold(
-              invalid => {
-                log.error(Json.stringify(JsError.toJson(invalid)))
-                None
-              },
-              book => Some(book))
-        }
-      }
-    }
+  private implicit val bookActorExecutionContext: ExecutionContext = actorSystem.dispatchers.lookup("akka.actor.book-lookup")
+  private implicit val bookActorTimeout: Timeout = 0.2.seconds
+  private implicit val magentoServiceImplicit = magentoService
 
-    lazy val lookedUpBook = {
-      lookup(isbn) map {
-        _ map { bookData =>
-          cache.add(isbn, bookData)
-          bookData.as[Book]
-        } orElse {
-          cache.add(isbn, JsNull)
-          None
-        }
-      }
-    }
+  def findByIsbn(isbn: String): Option[Book] = BookAgent.get(isbn) map { _.as[Book] }
+}
 
-    cache.get(isbn) flatMap {
-      _ map cachedBook getOrElse lookedUpBook
-    } recoverWith {
-      case NonFatal(e) => lookedUpBook
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+
+  def get(isbn: String)(implicit magentoService: MagentoService, executionContext: ExecutionContext): Option[JsValue] =
+
+    cache.get.get(isbn) match {
+      case Some(json) => Some(json)
+      case None =>
+          magentoService.findByIsbn(isbn) onComplete {
+            case Failure(e) => log.error("Magento lookup failed.", e)
+            case Success(None) => log.warn(s"Magento unable to find book for $isbn.")
+            case Success(Some(json: JsValue)) => cache alter { _ + (isbn -> json) }
+          }
+          None
```

Comment:
The first pattern matching is probably not necessary here since you are actually returning the result of `cache.get.get(isbn)` as is.
What about:
```
val bookJson = cache.get.get(isbn)
if(bookJson.isEmpty) {
    magentoService.findByIsbn(isbn) onComplete {
        case Failure(e) => log.error("Magento lookup failed.", e)
        case Success(None) => log.warn(s"Magento unable to find book for $isbn.")
        case Success(Some(json: JsValue)) => cache alter { _ + (isbn -> json) }
    }
}
bookJson
```